### ISSUE TITLE
1 cent issue

### ISSUE THREAD

**Issue by madc** [Sylius/Sylius#2328]
I just discovered, that sylius is tempering with prices.
Reproduce:
* Checkout the current sylius master branch
* Install with default setup
* Log into Backend and edit any product.
* Enter 4.10 into the price input field and save.

If you now have a look at the price of the product, it costs 4.09 Euro. Same goes for 75.35 (becomes 75.34). Most other values are saved as exacted. If you edit prices directly in the database, they stay correct until they are updated via backend.
I know, it's just a one cent, but nevertheless its a critical issue.

**Comment by pjedrzejewski** [comment]
I have never experienced this issue. Is it for some specific product configuration or all price field behave like this?

**Comment by madc** [comment]
No, nothing. I just checked out the master branch, did a sylius:install imported all Fixtures, took the first product and edited it. Of course i did all this, because i had this issue in another installation. Is there something modifying the product on update? Can you reproduce it?

**Comment by madc** [comment]
I'm investigating this right know. The issue seems to the the float to int conversion in the price setter in Sylius\Component\Core\Model\ProductVariant.

```php
public function setPrice($price)
{
    $this->price = (int)$price;
    return $this;
}
```

If i `var_dump($price, (int)$price)`, i get the following output:

```
float 410
int 409
```
Gave it a try with PHP 5.4.35, 5.4.36, 5.5.14 with always the same result. Here my test-script:
```
$test = 4.10;
$test2 = $test*100;
$test3 = (int)$test;
$test4 = (int)$test2;

var_dump($test, $test2, $test3, $test4);
```

Again the output:
```
float 4.10
float 410
int 4
int 409
```
Here is [a stack overflow post](http://stackoverflow.com/questions/5651026/surprising-float-int-conversion-in-php) with the same issue.

Seems to happen only with a couple of numbers like 4.10, 9.45, 75.35.

**Comment by inspiran** [comment]
I am on 5.6.4 and get the same results.

**Comment by madc** [comment]
To trace the issue a bit further, the transformation from 4.10 to 410 happens somewhere in the `Sylius\Bundle\MoneyBundle\Form\Type\MoneyType`. Here to be precise: 

```php
    public function setDefaultOptions(OptionsResolverInterface $resolver)
    {
        $resolver
            ->setDefaults(array(
                'currency' => $this->currency,
                'divisor'  => 100
            ))
        ;
    }
```
This converts the float correctly from 4.10 to 410, but it is technically still a float when it is given as a parameter to setPrice. IMHO this is a symfony bug based on a wrongful php calculation, but nevertheless we have to fix it somehow.

**Comment by madc** [comment]
A quick and (really) dirty fix would be to cast the $price to string first.

```
public function setPrice($price)
{
    $this->price = (int)(string)$price;
    return $this;
}
```

**Comment by madc** [comment]
Hm, i thought that there would be more discussion on this issue.

**Comment by alanpearce** [comment]
I ran into the same PHP issue when doing a similar conversion without using Sylius's helpers.
```php
(int) round($val * 100);
```
worked and conveys the intent quite well.

**Comment by madc** [comment]
@alanpearce is right.

```php
$test = 4.10;
$test2 = $test*100;
$test3 = (int)$test;
$test4 = (int)$test2;
$test5 = (int) round($test);
$test6 = (int) round($test2);

var_dump($test, $test2, $test3, $test4, $test5, $test6);
```
gives:
```
float 4.1
float 410
int 4
int 409
int 4
int 410
```

Would that be an acceptable solution, @pjedrzejewski?

**Comment by alanpearce** [comment]
Thought I'd check to see whether the behaviour changes with PHP version. [It doesn't](http://3v4l.org/MsMuU)

**Comment by michalmarcinkowski** [comment]
Here the problem is with Floating point limitations. Float value 4.1 is actually in decimal ~4.0999755859375

1. ~4.0999755859375
2. multiply by 100 = ~409.99755859375
3. cast to int (cut the float part) 4
4. cast to int (cut the float part) 409
5. round 4.10 (0.1 < 0.5 = 0) = 4
6. round 409.99755859375 (0.99... >= 0.5 = 1) = 409 + 1 = 410

@madc we will do a quick fix with proper rounding, but a target solution is to use a value object for money management (probably some existing library).

**Comment by pjedrzejewski** [comment]
Fixed in #2427.
